stackslib/src/chainstate/stacks/boot/pox-4.clar

```diff
@@ -1275,12 +1313,72 @@
           reward-set-indexes: (list),
           first-reward-cycle: first-reward-cycle,
           lock-period: lock-period,
-          delegated-to: (some tx-sender) })
+          delegated-to: (some tx-sender),
+          signing-key: next-signing-key })
 
       ;; return the lock-up information, so the node can actually carry out the lock.
       (ok { stacker: stacker,
             unlock-burn-height: new-unlock-ht }))))
 
+;; This function allows a node to register a new 'current-signer-set' during the prepare phase
+(define-private (set-current-signer-set (new-signer-set (list 4000 {signer: (buff 33)})))
+    (let 
+        (
+            (next-reward-cycle (+ u1 (current-pox-reward-cycle)))
+            (next-reward-cycle-burn-height (reward-cycle-to-burn-height next-reward-cycle))
+            (next-signer-set-map (map-get? reward-cycle-signing-key-ids next-reward-cycle))
+        )
+
+        ;; Assert not already set by checking for 'none' map entry
+        (asserts! (is-none next-signer-set-map) (err ERR_NEXT_SIGNER_SET_ALREADY_REGISTERED))
+
+        ;; Assert in prepare phase (height equal to / greater than start of prepare phase)
+        (asserts! (>= block-height (- next-reward-cycle-burn-height u100)) (err ERR_NOT_IN_PREPARE_PHASE_EARLY))
+
+        ;; Assert in prepare phase (height less than end of prepare phase)
+        (asserts! (< block-height next-reward-cycle-burn-height) (err ERR_NOT_IN_PREPARE_PHASE_LATE))
+
+        ;; Updates new signer-set
+        (var-set current-signer-set new-signer-set)
+    )
+)
+
+;; This function allows a node to register a list of key-ids to the current-signer-set
+;; This work is done one-at-a-time by the node & could require up to 4000 individual writes
+(define-private (set-current-signer (signer (buff 33)) (key-ids (list 4000 uint)))
```

Comment:
There is currently no test coverage for this.  Please add it to `pox_4_tests.rs`.  As before, test each error path as well as the happy path.